### DISCUSSION TITLE
Custom hook with priority

### DISCUSSION THREAD

**Discussion by nonamich** [vikejs/vike#1510] [Discussion]
👋🏼 Hello! First of all, thank you for a great tool.

Can I make my own hook that will be triggered on certain pages, on the server side, before calling the "+data" hook? I need to do some manipulations with the pageContext before getting to the "+data".

**Comment by brillout** [comment]
What is your use case?

(In general, yes, custom hooks allow you to do these kind of things.)

**Reply by nonamich** [reply]
In my case, some pages (in my example it's @topic) have different databases (don't ask why), but they have the same schemas, tables like posts, user, etc.

According to this documentation https://vike.dev/meta, everything works great, but the value returned in `pageContext.config` is exactly what `export default` returns. In my case I want the value to be the result of a function call to `pageContext`.

In short, I need a hook that works like `+onBeforeRender` but is called before the `+data` hook.

In general, this is not such a serious problem compared to next.js. I can solve it in other ways, but less concise

```
pages/
 @topic/
   +topic-db.ts (add db instance to pageContext if topic exists or throw render(404))
   @postId/
     +Page.tsx
     +data.ts (with the db instance that was added from +topic-db.ts)
     @SomeSubRoutes/** (also with this db instance)
```

**Reply by brillout** [reply]
I was thinking more of having one global `onBeforeRender()` hook that orchestrates two custom local hooks `onData()` and `onBeforeData()`.  Similarly to what is described at https://vike.dev/onBeforeRender#onbeforerender-meta.

I don't know the details so I'm limited in what I can recommend: feel free to elaborate what `+topic-db.ts` does.

Btw. your use case seems enterprise-ish while you're using unique Vike features: is [sponsoring](https://github.com/vikejs/vike/issues/1350) something your company would be up for?

**Reply by nonamich** [reply]
@brillout I'm using Vike for my own pet project, at this stage I don't have the ability to support Vike, but if you don't mind I can add something like "built with vike.dev"

Well, +topic-db.ts should do something like this: 
```typescript
const dbInstances = {
  blog1: new DB('localhost:5311'),
  blog2: new DB('localhost:5312'),
  blog3: new DB('localhost:5313'),
};

export const onBeforeData = (ctx: PageContextServer) => {
  const { blogName } = ctx.routeParams;
  const db = dbInstances[blogName];

  if (!db) {
    throw render(404);
  }

  return {
    pageContext: {
      db,
    },
  };
};

export const data = async ({ db }: PageContextServer) => {
  return { posts: await db.posts.findMany() };
};
```

In my dreams, custom hooks work like onBeforeRender, so that the database instance is available not only in `@blogName/+data`, but also in `@blogName/@postId/+data`, and so that there is no such property as [override](https://vike.dev/onBeforeRender#override) in onBeforeRender

**Reply by brillout** [reply]
I see. So, yea, I'd create two custom hooks then. AFAICT that should work, but let me know if it doesn't.

Adding further hooks like `onBeforeData()` is a consideration. Also, being able to make hooks cumulative is on the radar as well.
